File: 0040-combination-sum-ii/0040-combination-sum-ii.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function(candidates, target) {
    const results = []
    
    // sorting the candidates makes it easier to handle duplicate numbers (next to each other)
    candidates.sort((a, b) => a - b)
    
    const dp = (combination, remain, current, results) => {
        // base case: remain has reached target, so push copy of combination to final result
        if (remain === 0) {
            results.push(combination.slice())
            return
        }
        
        // initialize to -1 to prevent error on first iteration
        let prevCandidate = -1
        
        for (let i = current; i < candidates.length; i++) {
            const nextCandidate = candidates[i]
            
            // skip index when current candidate is same as previous
            if (nextCandidate === prevCandidate) {
                continue
            }
            
            // optimize: break out of loop if we pass target
            if (remain - nextCandidate < 0) break;
            
            // decision: take the nextCandidate
            combination.push(nextCandidate)
            dp(combination, remain - nextCandidate, i + 1, results)
            
            // decision: don't take nextCandidate (pop it back out before next iteration)
            combination.pop()
            
            // store the previous candidate
            prevCandidate = candidates[i]
        }
        
    }

    dp([], target, 0, results)
    
    return results;
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        # sort the candidates so it's easier to handle duplicates (next to each other)
        candidates.sort()
        results = []
        
        def backtrack(current_combination, current_index, remaining):
            # base case: we reached the target, so add a copy of the combination to result
            if remaining == 0:
                results.append(current_combination.copy())
                return
            
            # base case: reached passed the target, return
            if remaining < 0:
                return
            
            # mark previous_candidate as -1 (all candidates > 0) to avoid error on first iteration
            previous_candidate = -1
            
            # iterate from current position to rest of candidates
            for i in range(current_index, len(candidates)):
                next_candidate = candidates[i]
                
                # skip this iteration if duplicate candidate
                if previous_candidate == next_candidate:
                    continue;
                
                # decision: take the next_candidate
                current_combination.append(next_candidate)
                backtrack(current_combination, i + 1, remaining - next_candidate)
                
                # decision: don't take next_candidate (pop it out before next iteration)
                current_combination.pop()
                
                # mark previous candidate as the next_candidate
                previous_candidate = next_candidate
        
        backtrack([], 0, target)
        return results